docs/querying/nested-columns.md

```diff
@@ -0,0 +1,503 @@
+[["EXPR$0","EXPR$2","EXPR$3"],["COMPLEX<json>","STRING","STRING"],["OTHER","VARCHAR","VARCHAR"],["{\"x\":\"y\"}","\"{\\\"x\\\":\\\"y\\\"}\"","{\"x\":\"y\"}"]]
+```
+
+### Using helper operators
+
+The `JSON_KEYS` and `JSON_PATHS` functions are helper operators that you can use to examine JSON object schema. Use them to plan your queries, for example to work out which paths to use in `JSON_VALUE`.
+
+#### Example query: Examine JSON object schema
+
+ The following query illustrates how to use the helper operators to examine a nested data object.
+
+![Examine JSON object schema](../assets/nested-examine-schema.png)
+
+```sql
+SELECT
+  ARRAY_CONCAT_AGG(DISTINCT JSON_KEYS(shipTo, '$.')),
+  ARRAY_CONCAT_AGG(DISTINCT JSON_KEYS(shipTo, '$.address')),
+  ARRAY_CONCAT_AGG(DISTINCT JSON_PATHS(shipTo))
+FROM nested_data_example
+```
+
+Example query results:
+
+```json
+[["EXPR$0","EXPR$1","EXPR$2","EXPR$3"],["COMPLEX<json>","COMPLEX<json>","STRING","STRING"],["OTHER","OTHER","VARCHAR","VARCHAR"],["{\"x\":\"y\"}","\"{\\\"x\\\":\\\"y\\\"}\"","\"{\\\"x\\\":\\\"y\\\"}\"","{\"x\":\"y\"}"]]
+```
+
+## Known issues
+
+Before you start using the nested columns feature, consider the following known issues:
+
+- Directly using `COMPLEX<json>` typed columns and expressions is not well integrated into the Druid query engine. It can result in errors or undefined behavior when grouping and filtering, and when you use using `COMPLEX<json>` objects as inputs to aggregators. As a workaround, consider using `TO_JSON_STRING` to coerce the values to strings before you perform these operations.
+- Directly using array-typed outputs from `JSON_KEYS` and `JSON_PATHS` is moderately supported by the Druid query engine. You can group on these outputs, and there are a number of array expressions that can operate on these values, such as `ARRAY_CONCAT_AGG`. However, some operations are not well defined for use outside array-specific functions, such as filtering using `=` or `IS NULL`.
+- Input validation for JSON SQL operators is currently incomplete, which sometimes results in undefined behavior or unhelpful error messages.
+- Ingesting JSON columns with a very complex nested structure is potentially an expensive operation, and may require you to tune ingestion tasks and/or cluster parameters to account for increased memory usage or overall task run time. When you tune your ingestion configuration, treat each nested literal field inside a JSON object as a flattened top-level Druid column.
+
+## Further reading
+
+For more information, see the following pages:
+
+- [Nested columns functions reference](./sql-json-functions.md) for details of the functions used in the examples on this page.
+- [Ingestion spec reference](../ingestion/ingestion-spec.md) for information on native ingestion and [`transformSpec`](../ingestion/ingestion-spec.md#transformspec).
+- [Multi-stage query architecture overview](../multi-stage-query/index.md) for information on how to set up and use this feature.
```

Comment:
Suggest reordering to have `transformSpec` and `flattenSpec` details next to each other
```suggestion
- [Multi-stage query architecture overview](../multi-stage-query/index.md) for information on how to set up and use this feature.
- [Ingestion spec reference](../ingestion/ingestion-spec.md) for information on native ingestion and [`transformSpec`](../ingestion/ingestion-spec.md#transformspec).
```

Reply:
Done